Bookkeeper migration

rustyrussell opened the pull request:
This builds on #8445 

We do all the things:
1. We put the coin movement notifications into the db, in separate tables.
2. We implement listchainmoves and listchannelmoves (and wait commands) on top of this.
3. We implement sql plugin support for these tables, and optimize it.
4. We migrate the internal tables & annotations that bookkeeper does from its local db into the datastore.
5. We add new (internal, for now) interfaces to inject "foreign" utxo deposits and spends.
6. We enhance libplugin to allow sync commands at any time, to avoid rewriting everything!
7. We move bookkeeper off the notifications onto the list commands, so no more replays!
8. We convert bookkeeper to use the sql plugin commands to query the db.
9. We migrate bookkeeper's accounts.db into our internal db

It needs some more testing, and I'm sure it dosn't pass CI, but it's ready for review!

Closes: #8393

cdecker commented:
Cache cleared, restarting CI.

niftynei commented on tests/test_coinmoves.py:
```diff
@@ -1117,6 +1117,286 @@ def test_coinmoves_unilateral_htlc_dust(node_factory, bitcoind):
+    #                        'part_id': 0,
+    #                        'payment_hash': inv['payment_hash'],
+    #                        'primary_tag': 'invoice'}]
+    # expected_channel2 += [{'account_id': fundchannel['channel_id'],
```
this is a duplicate of the onchain event at L1324 that credits the account via an onchain movement?

niftynei reviewed (commented):
checkpoint submission #1

rustyrussell commented:
I've done some more work:
1. Minor fixups from @niftynei review (mainly correcting tests and comments). 
2. Rebased onto master now part 1 is merged.
3. Reworked final two commits (migrations) and implemented the full tests for them.

I appended the fixes needed, for later rebasing.

niftynei commented on plugins/bkpr/onchain_fee.c:
```diff
@@ -260,10 +286,10 @@ static void insert_chain_fees_diff(struct bkpr *bkpr,
 			continue;
 		if (!amount_msat_accumulate(&current_amt, ofs[i]->credit))
-			db_fatal(bkpr->db, "Overflow when adding onchain fees");
+			plugin_err(cmd->plugin, "Overflow when adding onchain fees");
```
the recorder tests pass in NULL for `cmd`, which would blow this up... guess we don't test for this :)

rustyrussell replied:
Yes, since this Shouldn't Happen.... since the result is supposed to be a different kind of crash, I don't think we care!

niftynei commented on plugins/bkpr/descriptions.c:
```diff
@@ -0,0 +1,323 @@
+	ud = utxo_desc_htable_get(descriptions->by_utxo, outpoint);
+	if (ud) {
+		tal_free(ud->desc);
+		ud->desc = tal_strdup(ud, desc);
```
we say we're `TAKES` the `desc` in the method params, but we strdup it here (and below)

rustyrussell replied:
tal_strdup() takes :)

niftynei commented on plugins/bkpr/descriptions.c:
```diff
@@ -0,0 +1,323 @@
+		struct bitcoin_outpoint outp;
+
+		keytok = json_get_member(buf, t, "key");
+		if (keytok->size != 4)
```
what case would this be true for? should we log/report on it?

niftynei commented on plugins/bkpr/descriptions.c:
```diff
@@ -0,0 +1,323 @@
+			continue;
+
+		/* ["bookkeeper", "description", "utxo", <key>] */
+		hextok = json_get_member(buf, t, "hex");
```
where is the `hex` keyword in the datastore struct that's returned? does every entry return a `key` and a `hex`?

rustyrussell replied:
Yes, we will always have it on the final key (there won't be any on directories).

niftynei commented on plugins/libplugin.c:
```diff
@@ -1568,25 +1603,16 @@ static struct command_result *handle_init(struct command *cmd,
 	/* FIXME: Move this to its own function so we can initialize at a
```
i think this fixme can be removed?

niftynei commented on plugins/bkpr/test/run-recorder.c:
```diff
@@ -1,1249 +0,0 @@
-	 * chan-1, 2222,  50
-	 * chan-1, 3333, 150
-	 * chan-1, 4444, 100
-	 */
```
these are pretty important tests to keep imo. it's not just database ops, it's internal fee allocation tests.

rustyrussell replied:
I agree, but I ran out of time :(

It will have to enter the events in an internal sqlite3 db and simulate the calls to the sql plugin on that...

rustyrussell replied:
OK, restored.  I'm not getting those hours of my life back tho!

rustyrussell commented on plugins/bkpr/db.c:
```diff
@@ -52,7 +52,7 @@ static struct migration db_migrations[] = {
 	{SQL("CREATE TABLE chain_events ("
 		"  id BIGSERIAL"
-		", account_id BIGINT REFERENCES accounts(id)"
+		", account_name TEXT"
```
It does seem that way, but these are not used (we never actually run the migrations in the end).

Actually creating the migrations to wean off the account indirection was just going to be work, so I cheated.
